servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/GrpcUtils.java

```diff
@@ -312,13 +316,35 @@ static <Resp> Publisher<Resp> validateResponseAndGetPayload(final StreamingHttpR
         validateContentType(headers, expectedContentType);
         final GrpcStatusCode grpcStatusCode = extractGrpcStatusCodeFromHeaders(headers);
         if (grpcStatusCode != null) {
+            // Drain the response messageBody to make sure concurrency controller marks the request as finished.
+            // In case the grpc-status is received in headers, we expect an empty messageBody, draining should not see
+            // any other frames. However, the messageBody won't complete until after the request stream completes too.
+            final Completable drainResponse = response.messageBody().beforeOnNext(frame -> {
+                throw new GrpcStatus(INTERNAL, null, "Violation of the protocol: received unexpected " +
+                        (frame instanceof HttpHeaders ? "Trailers" : "Data") +
+                        "frame after already receiving Trailers-Only response with grpc-status: " +
+                        grpcStatusCode.value() + '(' + grpcStatusCode + ')').asException();
+            }).ignoreElements();
             final GrpcStatusException grpcStatusException = convertToGrpcStatusException(grpcStatusCode, headers);
             if (grpcStatusException != null) {
-                // Give priority to the error if it happens, to allow delayed requests or streams to terminate.
-                return Publisher.<Resp>failed(grpcStatusException)
-                        .concat(response.messageBody().ignoreElements());
+                // In case of an error, we cannot concat GrpcStatusException after drainResponse because users may never
+                // see the exception if the request publisher never terminates, or they may see a TimeoutException that
+                // will hide the original exception. Therefore, we have to return an error asap and then immediately
+                // subscribe & cancel the drainResponse. Cancellation is necessary to prevent sending large request
+                // payloads over network when server returns an error.
+                return Publisher.<Resp>failed(grpcStatusException).afterOnError(__ -> {
+                    // Because we subscribe asynchronously, users won't receive any further errors from drainResponse.
+                    // Instead, we log those errors for visibility. Use onErrorComplete instead of whenOnError to avoid
+                    // logging the same exception twice inside SimpleCompletableSubscriber.
+                    drainResponse.onErrorComplete(t -> {
+                        LOGGER.error("Unexpected error", t);
```

Comment:
consider adding some more context in the error msg, eg. mentioning that it happened when trailers-only were received.